cpu/qn908x: Initial minimal support for NXP QN908x CPUs.

iosabi opened the pull request:
### Contribution description

The NXP QN908x CPU family is a Cortex-M4F CPU with integrated USB,
Bluetooth Low Energy and in some variants NFC. This patch implements the
first steps for having support for this CPU.

While the QN908x can be considered the successor of similar chips from
NXP like the KW41Z when looking at the feature set, the internal
architecture, boot image format and CPU peripherals don't match those
in the Kinetis line. Therefore, this patch creates a new directory for
just the QN908x chip under cpu/qn908x.

The minimal set of peripherals are implemented in this patch to allow
the device to boot and enable a GPIO: the gpio and wdt peripheral
modules only.

The wdt driver is required to boot and disable the wdt. On reset, the
wdt is disabled by the chip, however the QN908x bootloader stored in
the internal ROM enables the wdt and sets a timer to reboot after 10
seconds, therefore it is needed to disable the wdt in RIOT OS soon
after booting. This patch sets it up such that when no periph_wdt module
is used the Watchdog is disabled, but if the periph_wdt is used it must
be configured (initialized) within the first 10 seconds.

The header files under the vendor/ directories (both
cpu/qn908x/include/vendor/ and cpu/qn908x/vendor/) are part of NXP's
SDK for the QN908x family available for download from:
  https://mcuxpresso.nxp.com/en/builder
The files included in this vendor/ directory are released by NXP under
an Open Source license as described in each file, but only the files
used by this patch are included here.

### Testing procedure

Defined a custom board for this CPU and compiled a simple application
that blinks some LEDs. Manually tested with periph_wdt and with
periph_wdt_cb as well.

### Issues/PRs references

This is part of the FR #13852 . This patch depends on PR #13851 (among other things) to create a valid image but they can be merged independently.

iosabi commented on cpu/qn908x/Makefile.dep:
```diff
@@ -0,0 +1,16 @@
+# Periph modules in this CPU are defined whenever possible based on the vendor
```
This is something up for discussion. Many of the headers in the vendor/ directory define useful constants and inline functions that can be used from drivers for RIOT-OS, but sometimes the intended way to use the module in the vendor SDK is different than the standard API in RIOT-OS. Take the example of WDT in this commit, the vendor's fsl_wdt.c file was not very useful (initialization of the WDT in a different way) and therefore not included here. In other cases, like the fsl_clock.c it is really useful to have the constants like `kCLOCK_Wdt` defined in fsl_clock.h and the CLOCK_GetFreq() function which does all the proper math to figure out the current frequency of a given clock source. I think that the overall goal of reuse these modules with other variants might not be possible so I'd appreciate some guidance on whether I should try to use vendor functions from the vendor SDK when possible or just access the registers directly with the headers defined in QN908XC.h.

iosabi replied:
I ended up removing most of the vendor SDK files in the vendor/drivers directory as they are not that useful and other drivers don't match the API in RIOT-OS so a re-implementation would be needed for parts of them. Only fsl_clock and fsl_iocon (the mux) are there now.

benpicco commented:
The code looks good.
Is there an evaluation board with this CPU or can you add the board you are using?

It should be possible for CI to compile-test this code.

iosabi commented:
There is an evaluation board from NXP, the [QN9080DK board](https://www.nxp.com/products/wireless/bluetooth-low-energy/a-highly-extensible-platform-for-application-development-of-qn908x:QN9080DK), however I didn't buy that board because its selection of peripherals didn't match want I needed. I'm testing with a custom board I made with a QN9080. This is not really an issue since there is a lot of support needed in the cpu module for things that are independent of the board (like the uart, i2c, wdt, timers, etc). I plan to send commits for a "common" board with some things you need regardless of your board layout that I think need to go in the board module and after that I could create a qn9080dk board module with the board definition of the developer board since the schematics of the evaluation board are public, but I wouldn't be able to test that part on hardware. However in the interest of sending small and easy to review commits I'm only including the initial support here. Should I send more commits together instead?

I'll push an update to this patch to fix an issue with the gpio irq (I forgot to enable it in the NVIC).

benpicco commented:
> Should I send more commits together instead?

I'm all for small PRs - but I'd like to have a simple board so the code can run through CI.
Even a dummy board would be fine (I see that QN9080 is also sold a module, so that would probably even be useful).

benpicco commented:
@iosabi I'd still like to have this in - but we need a board (of any kind) in order to compile this, so it doesn't bit-rot away.

iosabi commented:
@benpicco Hey, sorry for being unresponsive with this for so long. The current situation didn't leave me much time to work on this but I'll dedicate more time to this now.

I updated the patches to work on top of master and added a Kconfig. This pull request now has 3 commits:

* one for the cpu/qn908x
* one for the boards/common/qn908x (there's a fair amount of things to deal with at the board level regardless of the actual board you use)
* a new one for boards/qn9080dk for the base board of the development kit.

This being the minimum possible to make things compile/work together it only implements GPIO and WDT. I have more commits in the pipeline to add UART, I2C and SPI, but I will add those as follow up commits since this is already large.

Would this be enough for the CI to run? I tested it with:
```
make BOARD=qn9080dk -C examples/hello-world/
```
(and the examples/saul too)

Regarding the choice of board, there's a "QN9080-001-M17Z" "module" (it looks like epoxied together so somewhere between a board and stacked chip) but that one is for the QN9080SIP which includes NFC. I don't have one of those and I'm not sure if it makes more sense to handle that little board as a cpu_model or as a board. Anyway, the qn9080dk fits better the description of a board and would be a better example if you want to make your own board based on this or stack a module on top of the DK GPIOs.

Finally, I'm getting this error in the `make static-test` but I don't know how to solve it, other cortex MCUs have the same code:
```
Running "./dist/tools/cppcheck/check.sh" x
Command output:

	cpu/qn908x/vectors.c:30: error (unknownMacro): There is an unknown macro here somewhere. Configuration is required. If ISR_VECTOR is a macro then please configure it.
```

benpicco commented:
> Finally, I'm getting this error in the `make static-test` but I don't know how to solve it, other cortex MCUs have the same code:
> 
> ```
> Running "./dist/tools/cppcheck/check.sh" x
> Command output:
> 
> 	cpu/qn908x/vectors.c:30: error (unknownMacro): There is an unknown macro here somewhere. Configuration is required. If ISR_VECTOR is a macro then please configure it.
> ```

Not sure how you got those, [CI](https://ci.riot-os.org/) seems to be happy with the static tests.
Only [`tests/kconfig_features`](https://ci.riot-os.org/RIOT-OS/RIOT/13855/15941a723e7965cec5702b2dfd247a960119f43e/output/compile/tests/kconfig_features/qn9080dk:gnu.txt) fails which means there is a discrepancy between the options selected by Kconfig and the old build system.

iosabi commented:
I'm looking at the tests/kconfig_features failure, I'll update this with a fix.

benpicco reviewed (commented):
looks pretty good.
Please make a separate commit for adding the vendor files, that makes the review easier.

iosabi commented on boards/qn9080dk/Kconfig:
```diff
@@ -0,0 +1,19 @@
+
+    # Put defined MCU peripherals here (in alphabetical order)
+    select HAS_PERIPH_GPIO
+    select HAS_PERIPH_GPIO_IRQ
```
I was actually wondering that. The CPU always has the periph_gpio module, but whether those are used as gpios or something else depends on the wiring. If the board has no external buttons/gpio/leds etc, would a board with this cpu provide the PERIPH_GPIO feature?

how about other periphs? (uart, i2c, spi)

benpicco replied:
>  If the board has no external buttons/gpio/leds etc, would a board with this cpu provide the PERIPH_GPIO feature?

That's a bit of a philosophical question almost ;)
The thing is that a uart or spi implementation might configure GPIOs (e.g. chip select) and GPIOs don't require any configuration, they are always there.
So it doesn't hurt to always advertise GPIO support by the CPU to keep the board code shorter. It's only used if something `FEATURES_REQUIRED += periph_gpio` anyway.

> how about other periphs? (uart, i2c, spi)

￼Those require some configuration (which pins to use) and they would need to be exposed in order to be useful, so it can only be advertised by the board if they are configured.

iosabi replied:
I'm thinking that an application that implements a blinking LED will set `FEATURES_REQUIRED += periph_gpio` and maybe use them from saul, but it is possible to not compile and include gpio.c in the hello-world application case, even after implementing a uart, i2c and spi drivers. The CS is handled by hardware here too and doesn't involve calling gpio.c code (it does involve setting the mux accordingly, but it is not a GPIO). All of this to say that maybe a board can save some space by not including gpio.c, but it doesn't hurt too much to have it. I moved this to the cpu/.

benpicco replied:
`FEATURES_PROVIDED` is only a hint to the build system that the module*can* be selected. 
It will only be compiled in if it is also in `FEATURES_REQUIRED`

iosabi commented on boards/qn9080dk/board.c:
```diff
@@ -0,0 +1,73 @@
+    cpu_init();
+}
+
+void clocks_init(void)
+{
+    /* Set up clock selectors - Attach clocks to the peripheries */
+
+    /* Switch XTAL_CLK to 32M */
+    CLOCK_AttachClk(k32M_to_XTAL_CLK);
+    /* Switch 32K_CLK to XTAL32K */
+    CLOCK_AttachClk(kXTAL32K_to_32K_CLK);
+    /* Switch SYS_CLK to XTAL */
+    CLOCK_AttachClk(kXTAL_to_SYS_CLK);
+    /* Switch WDT_CLK to APB */
+    CLOCK_AttachClk(kAPB_to_WDT_CLK);
+
+    /* Set up dividers */
+
+    /* Set OSC32M_DIV divider to value 2 */
+    CLOCK_SetClkDiv(kCLOCK_DivOsc32mClk, 1U);
+    /* Set XTAL_DIV divider to value 2 */
+    CLOCK_SetClkDiv(kCLOCK_DivXtalClk, 1U);
+    /* Set AHB_DIV divider to value 2 */
+    CLOCK_SetClkDiv(kCLOCK_DivAhbClk, 1U);
+    /* Set FRG_MULT1 to value 0, Set FRG_DIV1 to value 255 */
+    CLOCK_SetClkDiv(kCLOCK_DivFrg1, 0U);
+    /* Set FRG_MULT0 to value 0, Set FRG_DIV0 to value 255 */
+    CLOCK_SetClkDiv(kCLOCK_DivFrg0, 0U);
+    /* Set APB_DIV divider to value 1 */
+    CLOCK_SetClkDiv(kCLOCK_DivApbClk, 0U);
+}
```
ok, the thinking here is that whether you have an external XTAL for 32 MHz (or 16 MHz) and another external xtal for the 32.768 KHz depends on the board wiring, you can have a design that uses only the internal ones (32MHz and 32.768 KHz), then based on this you would have different clock setup for the board which also changes how do you configure the APB bus clock for the peripherals.

Should all of this be done using macros in `cpu_conf.h` that the board can override, but the code lives in cpu_init() (or a function called from there)?

benpicco replied:
> Should all of this be done using macros in cpu_conf.h that the board can override, but the code lives in cpu_init()

Yes that's the idea.
It's better to have all the logic in one place and not spread out / copied across multiple boards.

You could have a `BOARD_HAS_XOSC32K` define that indicates if a 32kHz crystal is present, and another `BOARD_XOSC_FREQ` / `BOARD_HAS_XOSC` define to indicate the frequency / presence of a high frequency oscillator.

There are no naming conventions for this, just use whatever best matches the description in the data sheet.

iosabi replied:
I moved this to cpu.c and added a whole lot of config.

I'm not entirely sure how is the Kconfig supposed to work. If I run `make BOARD=qn9080dk -C examples/hello-world/` the `autoconf.h` is not generated and the CONFIG_ variables with the defaults in the Kconfig are not set, not even the ones the board should be forcing.

I ended up adding the same logic with CONFIG_* variables in the cpu_conf.h, maybe this is just a transition thing but it would be better if I don't have to do that. On the other hand adding them to cpu_conf.h lets me document them so doxygen can pick them up.

benpicco replied:
Yes the Kconfig transition is not yet complete, not all modules can be selected in Kconfig yet, so it's not enabled by default.

I *think* you don't need to model the board oscillator config in Kconfig since it's soldered in place.

iosabi replied:
I added the matching "CONFIG_*" to the CFLAGS in `Makefile.include` in the board, so the if you don't have an autoconf.h you still get the CONFIG_BOARD_HAS_XTAL_32M and related macros defined.

iosabi commented on cpu/qn908x/periph/mux_common.c:
```diff
@@ -0,0 +1,32 @@
+
+#include "vendor/drivers/fsl_iocon.h"
+
+void mux_set_pin(gpio_t pin, uint32_t func)
```
The issue with this is that is not gpio related only. The iocon is used to select the function of the pin, which "GPIO" is one of those functions. I'll use this from other drivers to set the I/O pin to the right function.

benpicco replied:
Yes that's the case for the other implementations as well.
Don't worry about `gpio.c` being compiled even though most functions are unused - we enable `-ffunction-sections` and `--gc-sections` so unused functions are thrown out by the linker.

iosabi replied:
I see that other CPUs have "gpio_init_mux()" declared in the periph_cpu.h. I renamed this to "gpio_init_mux" and renamed this header to gpio_mux.h, but the implementation is not in gpio.c. I don't think it is a good idea to have the prototype in one place and the definition in another file, nor I think we should expose this to anybody importing periph_cpu.h.

Conceptually this is a different block in the MCU, it doesn't need the GPIO clock to be enabled (yeah, the GPIO block has a clock enable bit, don't ask me why but it is needed for the gpio to work, I guess you can save \epsilon power there if you never run gpio_init()), this is about multiplexing physical "pins", all of which support in particular GPIO.

iosabi commented on cpu/qn908x/periph/gpio.c:
```diff
@@ -0,0 +1,276 @@
+         */
+        base->INTSTATUS = status;
+        while (status) {
+            uint32_t pin = bitarithm_lsb(status);
+            status ^= 1u << pin;
```
oh, good to know. I updated it.

iosabi commented on boards/common/qn908x/Makefile.dep:
```diff
@@ -0,0 +1,3 @@
+ifneq (,$(filter saul_default,$(USEMODULE)))
+  USEMODULE += saul_gpio
+endif
```
Done.

iosabi commented on boards/common/qn908x/Makefile.features:
```diff
@@ -0,0 +1,5 @@
+CPU ?= qn908x
+
+# Put defined MCU peripherals here (in alphabetical order)
+FEATURES_PROVIDED += periph_cpuid
+FEATURES_PROVIDED += periph_wdt periph_wdt_cb
```
Done.

iosabi commented:
I updated the PR, split the vendor files to a new commit (first in the stack) and added new Kconfig with the clock configuration. Please take a look again, thanks!

leandrolanzieri commented on cpu/qn908x/Kconfig.clk:
```diff
@@ -0,0 +1,115 @@
+    depends on CPU_FAM_QN908X
+
+config BOARD_HAS_XTAL32K
+    bool "Board has an external 32.768 KHz crystal"
```
I think this symbol should not have a prompt, as it is selected and used as interface from the board to indicate that it has a crystal:
```suggestion
    bool
```

iosabi replied:
Thanks, indeed this makes a lot of sense.

leandrolanzieri commented on cpu/qn908x/Kconfig.clk:
```diff
@@ -0,0 +1,115 @@
+endchoice
+
+config BOARD_HAS_XTAL
+    bool "Board has an external 32 MHz or 16 MHz crystal"
```
Same here:
```suggestion
    bool
```

iosabi replied:
Done.

iosabi replied:
I changed the BOARD_HAS_XTAL to be implied by BOARD_HAS_XTAL_32M or BOARD_HAS_XTAL_16M and made all three have no prompt. The xtal is certainly soldered to the board so it is part of the board definition to tell which one is there.

leandrolanzieri reviewed (commented):
@iosabi thanks for adding the clock configurations with Kconfig, I left some comments on that.

iosabi commented on boards/common/qn908x/Kconfig:
```diff
@@ -0,0 +1,10 @@
+# General Public License v2.1. See the file LICENSE in the top level
+# directory for more details.
+
+config BOARD_COMMON_QN908X
+    bool
+    select CPU_FAM_QN908X
+    # Add common board support here.
```
I replaced it with a "depends on" to make sure this is only used for qn908x boards, probably not needed.

leandrolanzieri replied:
Makes sense, but now that this symbol selects nothing, what would be the point on having it? Is there a plan to extend the features provided by the this common grouping?

iosabi replied:
I'm planning to eventually have RIOTBOOT which I think would be here, right?

leandrolanzieri replied:
Yes that's fine. Thanks for clarifying !

iosabi commented on boards/qn9080dk/Kconfig:
```diff
@@ -0,0 +1,22 @@
+
+    # Put defined MCU peripherals here (in alphabetical order)
+    select BOARD_HAS_XTAL
+    #select BOARD_XTAL
```
yeah, I actually wanted BOARD_XTAL_32M here, but it wasn't possible because it was a choice. I made those now a regular config so you can set either one from the board.

benpicco commented on cpu/qn908x/include/cpu_conf.h:
```diff
@@ -0,0 +1,328 @@
+ * Makefile.
+ */
+#define CPU_FLASH_BASE          (QN908X_ROM_START_ADDR)
+/** @} */
```
You can also set

```C
#define CPU_HAS_BITBAND 1
```

here

![image](https://user-images.githubusercontent.com/1301112/100749550-16530900-33e5-11eb-915d-7609e1ac902a.png)

iosabi replied:
Oh thanks, I missed that part.

benpicco commented:
This looks pretty good! Btw, if you have a debugger that supports it, you can also try `stdio_semihosting` to get some output from the board without UART.

benpicco commented on cpu/qn908x/include/vendor/QN908XC.h:
```diff
@@ -0,0 +1,9484 @@
+
+#ifndef _QN908XC_H_
+#define _QN908XC_H_                              /**< Symbol preventing repeated inclusion */
+
```
I hate modifying vendor files just as much as the next guy, but those need a

```C
#ifdef __cplusplus
extern "C" {
#endif

…

#ifdef __cplusplus
}
#endif
```

header guard.
[CI is complaining](https://github.com/RIOT-OS/RIOT/pull/13855/checks?check_run_id=1483161312) - usually vendor files are excluded from CI checks, but I see the point since they might transitively end up getting included in a global header file.

It's only 5 files, so I guess this is the easiest way.

iosabi replied:
Well, the thing is that the files that need this have it around the functions already. If you only have #define or typedef you don't need this extern "C" header guard, which is the case for those files. I can add them but it would be just to calm down externc/check.sh. I can deal with providing some tooling to update the vendor files that patch this in when needed. Vendor SDK updates are not that common anyway.

iosabi replied:
Vendor files also use "#if defined(__cplusplus)" but the check expects "#ifdef __cplusplus"

benpicco replied:
> Vendor files also use "#if defined(__cplusplus)" but the check expects "#ifdef __cplusplus"

urgh, yea `dist/tools/externc/check.sh` doesn't check for this - but this is a bug in the script.
If you could fix that, that would be great!
Not sure if there is a good way to determine if a header file only contains defines.

iosabi commented:
> This looks pretty good! Btw, if you have a debugger that supports it, you can also try `stdio_semihosting` to get some output from the board without UART.

I have already a UART driver for this using the FLEXCOMM modules which works fine, I'll send that patch once this is merged. The stdio_null was the easiest way for me to break the dependency on stdio and make examples compile. I've tested a larger stack but I think it is better to split things up for review.

Let me know if there's anything else to do here.

benpicco reviewed (approved):
CI is how happy too - looking forward to the follow ups :)

benpicco commented:
If you have UART you can also run `auto_test` from `tests/periph_gpio` to see if interrupts are working as expected. 

You'll need a timer, until you have a periph timer implemented #14553 would also do.

iosabi commented:
> If you have UART you can also run `auto_test` from `tests/periph_gpio` to see if interrupts are working as expected.
> 
> You'll need a timer, until you have a periph timer implemented #14553 would also do.

I have a timer in the pipeline... but timers need a bit more work because this cpu exposes a SCTimer/PWM combined block which you can in theory use for both timers and PWM, as well as a CTIMER block (four of them) which you can use for timer. I'll have xtimer working soon, but first I'll send you the RTC module which is simpler.

Regarding the GPIO interrupts I tested them manually just poking them :-)
